File: src/main/java/com/ririv/quickoutline/state/CurrentFileState.java
package com.ririv.quickoutline.state;

import com.google.inject.Inject;
import com.ririv.quickoutline.exception.EncryptedPdfException;
import com.ririv.quickoutline.service.PdfOutlineService;
import javafx.beans.binding.Bindings;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.ReadOnlyObjectWrapper;
import javafx.beans.property.SimpleObjectProperty;

import java.io.IOException;
import java.nio.file.Path;

public class CurrentFileState {

    private final ObjectProperty<Path> srcFileProperty = new SimpleObjectProperty<>();
    private final ReadOnlyObjectWrapper<Path> destFileWrapper;

    private final PdfOutlineService pdfOutlineService;

    @Inject
    public CurrentFileState(PdfOutlineService pdfOutlineService) {
        this.pdfOutlineService = pdfOutlineService;
        destFileWrapper = new ReadOnlyObjectWrapper<>();
        destFileWrapper.bind(Bindings.createObjectBinding(() -> {
            Path src = srcFileProperty.get();
            if (src == null) {
                return null;
            }
            return calculateDestFilePath(src);
        }, srcFileProperty));
    }

    public ObjectProperty<Path> srcFileProperty() {
        return srcFileProperty;
    }

    public ReadOnlyObjectProperty<Path> getDestFileProperty() {
        return destFileWrapper.getReadOnlyProperty();
    }

    public Path getSrcFile() {
        return srcFileProperty.get();
    }

    public Path getDestFile() {
        return destFileWrapper.get();
    }

    public void setSrcFile(Path file) throws IOException, EncryptedPdfException, com.itextpdf.io.exceptions.IOException {
        if (file == null) {
            clear();
            return;
        }
        // Centralized validation
        pdfOutlineService.checkOpenFile(file.toString());
        this.srcFileProperty.set(file);
    }

    public Path calculateDestFilePath(Path srcFilePath) {
        if (srcFilePath == null) {
            return null;
        }
        String srcFileName = srcFilePath.getFileName().toString();
        int dotIndex = srcFileName.lastIndexOf(".");
        String nameWithoutExt = (dotIndex == -1) ? srcFileName : srcFileName.substring(0, dotIndex);
        String ext = (dotIndex == -1) ? "" : srcFileName.substring(dotIndex);

        Path parentDir = srcFilePath.getParent();
        String destFileName = nameWithoutExt + "_new" + ext;
        return parentDir.resolve(destFileName);
    }

    public void clear() {
        this.srcFileProperty.set(null);
    }
}

File: src/main/kotlin/com/ririv/quickoutline/state/CurrentFileState.kt
package com.ririv.quickoutline.state

import com.ririv.quickoutline.exception.EncryptedPdfException
import com.ririv.quickoutline.service.PdfOutlineService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.IOException
import java.nio.file.Path

class CurrentFileState(private val pdfOutlineService: PdfOutlineService) {

    private val _srcFile = MutableStateFlow<Path?>(null)
    val srcFile = _srcFile.asStateFlow()

    private val _destFile = MutableStateFlow<Path?>(null)
    val destFile = _destFile.asStateFlow()

    fun setSrcFile(file: Path?) {
        if (file == null) {
            clear()
            return
        }
        // Centralized validation
        try {
            pdfOutlineService.checkOpenFile(file.toString())
            _srcFile.value = file
            _destFile.value = calculateDestFilePath(file)
        } catch (e: IOException) {
            // handle exception
        } catch (e: EncryptedPdfException) {
            // handle exception
        } catch (e: com.itextpdf.io.exceptions.IOException) {
            // handle exception
        }
    }

    private fun calculateDestFilePath(srcFilePath: Path?): Path? {
        if (srcFilePath == null) {
            return null
        }
        val srcFileName = srcFilePath.fileName.toString()
        val dotIndex = srcFileName.lastIndexOf(".")
        val nameWithoutExt = if (dotIndex == -1) srcFileName else srcFileName.substring(0, dotIndex)
        val ext = if (dotIndex == -1) "" else srcFileName.substring(dotIndex)

        val parentDir = srcFilePath.parent
        val destFileName = nameWithoutExt + "_new" + ext
        return parentDir.resolve(destFileName)
    }

    fun clear() {
        _srcFile.value = null
        _destFile.value = null
    }
}
